File: org.ai4fm.proofprocess.isabelle.core/src/org/ai4fm/proofprocess/isabelle/core/parse/SessionTracker.java
package org.ai4fm.proofprocess.isabelle.core.parse;

import java.util.EnumSet;
import java.util.Queue;
import java.util.Set;
import java.util.concurrent.ConcurrentLinkedQueue;

import org.ai4fm.proofprocess.isabelle.core.IsabelleProofPlugin;
import org.ai4fm.proofprocess.isabelle.core.analyse.ProofAnalyzer;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.IStatus;
import org.eclipse.core.runtime.Status;
import org.eclipse.core.runtime.jobs.IJobChangeEvent;
import org.eclipse.core.runtime.jobs.Job;
import org.eclipse.core.runtime.jobs.JobChangeAdapter;

import scala.collection.JavaConversions;

import isabelle.Command;
import isabelle.Document;
import isabelle.Session;
import isabelle.Session.Commands_Changed;
import isabelle.eclipse.core.IsabelleCorePlugin;
import isabelle.eclipse.core.util.SafeSessionActor;
import isabelle.eclipse.util.SessionEventSupport;
import isabelle.scala.ISessionCommandsListener;
import isabelle.scala.SessionActor;
import isabelle.scala.SessionEventType;

public class SessionTracker {

	private final SessionEventSupport sessionEvents;
	
	private final Job analysisJob;
	
	/**
	 * A concurrent queue is used for pending events, because the queue is
	 * likely to be updated from multiple threads.
	 */
	private final Queue<CommandAnalysisEvent> pendingEvents = 
			new ConcurrentLinkedQueue<CommandAnalysisEvent>();
	
	public SessionTracker() {
		
		this.analysisJob = new AnalysisJob();
		// add a listener that reschedules the job after completion if there are pending events
		analysisJob.addJobChangeListener(new JobChangeAdapter() {
			@Override
			public void done(IJobChangeEvent event) {
				if (!pendingEvents.isEmpty()) {
					// there are pending events, reschedule the job to consume them
					analysisJob.schedule();
				}
			}
		});
		
		this.sessionEvents = new SessionEventSupport(EnumSet.of(SessionEventType.COMMAND)) {
			
			@Override
			protected SessionActor createSessionActor(Session session) {
				return new SafeSessionActor().commandsChanged(new ISessionCommandsListener() {
					@Override
					public void commandsChanged(Commands_Changed changed) {
						addPendingAnalysis(changed);
					}
				});
			}
		};
		
		// TODO add a pending analysis for commands possibly already in the snapshot
	}
	
	public void dispose() {
		this.sessionEvents.dispose();
	}
	
	private void addPendingAnalysis(Commands_Changed changed) {
		
		Session session = IsabelleCorePlugin.getIsabelle().getSession();
		// Isabelle State is immutable, so just take the whole state for processing 
		// we can analyse what is needed in a low-priority analysis thread
		pendingEvents.add(new CommandAnalysisEvent(
				JavaConversions.setAsJavaSet(changed.commands()),
				session.current_state()));
		
		// wake up the analysis job
		analysisJob.schedule();
	}
	
	private static class CommandAnalysisEvent {

		private final Set<Command> changedCommands;
		private final Document.State docState;

		public CommandAnalysisEvent(Set<Command> changedCommands, Document.State docState) {
			super();
			this.changedCommands = changedCommands;
			this.docState = docState;
		}
	}
	
	private class AnalysisJob extends Job {

		public AnalysisJob() {
			super("Analysing proof process");
			// lowest priority
			setPriority(DECORATE);
		}

		@Override
		protected IStatus run(IProgressMonitor monitor) {
			
			CommandAnalysisEvent nextEvent = pendingEvents.poll();
			if (nextEvent == null) {
				// nothing to execute
				return Status.OK_STATUS;
			}
			
			if (IsabelleProofPlugin.getDefault() == null) {
				// shutting down - do not execute any more
				return Status.CANCEL_STATUS;
			}
			
			try {
				return analyze(nextEvent, monitor);
			} catch (CoreException e) {
				return IsabelleProofPlugin.error(
						"Proof process analysis failed: " + e.getMessage(), e);
			}
		}
	}
	
	private IStatus analyze(CommandAnalysisEvent event, IProgressMonitor monitor) throws CoreException {
		
		// delegate to the proof analyzer
		ProofAnalyzer proofAnalyzer = new ProofAnalyzer();
		return proofAnalyzer.analyze(event.changedCommands, event.docState, monitor);
	}
	
}

File: org.ai4fm.proofprocess.isabelle.core/src/org/ai4fm/proofprocess/isabelle/core/parse/SessionTracker.scala
package org.ai4fm.proofprocess.isabelle.core.parse

import java.util.Queue
import java.util.concurrent.ConcurrentLinkedQueue

import scala.actors.Actor._

import org.ai4fm.proofprocess.core.parse.TrackingToggle
import org.ai4fm.proofprocess.isabelle.core.IsabellePProcessCorePlugin.error
import org.ai4fm.proofprocess.isabelle.core.analysis.ProofAnalyzer
import org.ai4fm.proofprocess.isabelle.core.patch.IsabellePatcher
import org.eclipse.core.runtime.{CoreException, IProgressMonitor, IStatus, Status}
import org.eclipse.core.runtime.jobs.{IJobChangeEvent, Job, JobChangeAdapter}

import isabelle.{Command, Document, Session}
import isabelle.eclipse.core.IsabelleCore
import isabelle.eclipse.core.util.{LoggingActor, SessionEvents}


/**
  * @author Andrius Velykis 
  */
class SessionTracker extends SessionEvents {

  // the actor to react to session events
  // make it lazy, otherwise there are problems with plug-in loading sequence
  override protected lazy val sessionActor = LoggingActor {
    loop {
      react {
        case changed: Session.Commands_Changed => addPendingAnalysis(changed)
        case _ => // do nothing
      }
    }
  }

  // subscribe to commands change session events
  override protected def sessionEvents(session: Session) = List(session.commands_changed)

  /** Reacts to preference changes whether ProofProcess tracking is enabled */
  private val tracking = TrackingToggle { isTracking =>
    if (!isTracking) {
      // no longer tracking, so discard all pending events
      pendingEvents.clear
    }
  }

  def init() {
    initSessionEvents()
    tracking.init()
  }
  def dispose() {
    tracking.dispose()
    disposeSessionEvents()
  }

  /** A concurrent queue is used for pending events, because the queue is
    * likely to be updated from multiple threads.
    */
  private val pendingEvents = new ConcurrentLinkedQueue[CommandAnalysisEvent]

  private val analysisJob = new AnalysisJob
  // add a listener that reschedules the job after completion if there are pending events
  analysisJob.addJobChangeListener(jobDoneListener {
    if (!pendingEvents.isEmpty) {
      // there are pending events, reschedule the job to consume them
      analysisJob.schedule
    }
  })

  private val patcher = new IsabellePatcher


  private def addPendingAnalysis(changed: Session.Commands_Changed) = if (tracking.isTracking) {

    IsabelleCore.isabelle.session foreach { session =>
      
      // Isabelle State is immutable, so just take the whole state for processing 
      // we can analyse what is needed in a low-priority analysis thread
      pendingEvents.add(CommandAnalysisEvent(changed.commands, session.current_state))

      // wake up the analysis job
      analysisJob.schedule()
    }
  }


  /**
   * Retrieves all pending events and merges them.
   * 
   * Uses the last available snapshot state and collects all commands into one
   */
  private def allPendingEvents(): Option[CommandAnalysisEvent] = {
    val allPending = pollAll(pendingEvents)

    if (allPending.isEmpty) {
      None
    } else {
      val allCmdLists = allPending map (_.changedCommands)
      val allCmds = allCmdLists.flatten.toSet

      val lastState = allPending.last.docState
      Some(CommandAnalysisEvent(allCmds, lastState))
    }
  }

  private def pollAll[A](queue: Queue[A]): List[A] = Option(queue.poll()) match {
    case None => Nil
    case Some(elem) => elem :: pollAll(queue)
  }


  private class AnalysisJob extends Job("Analysing proof process") {

    // lowest priority
    setPriority(Job.DECORATE)

    override protected def run(monitor: IProgressMonitor): IStatus = {

      // merge pending events for performance improvements
//      val nextEvent = Option(pendingEvents.poll())
      val nextEvent = allPendingEvents()
      nextEvent match {
        // nothing to execute
        case None => Status.OK_STATUS
        case Some(event) => try {
          analyze(event, monitor)
        } catch {
          case e: CoreException => error(Some(e), Some("Proof process analysis failed: " + e.getMessage))
        }
      }
    }
  }

  private def jobDoneListener(f: => Unit) =
    new JobChangeAdapter {
      override def done(event: IJobChangeEvent) = f
    }

  @throws(classOf[CoreException])
  private def analyze(event: CommandAnalysisEvent, monitor: IProgressMonitor): IStatus =
    if (patcher.checkIsabellePatched()) {

      // delegate to the proof analyzer
      ProofAnalyzer.analyze(event.docState, event.changedCommands, monitor)

    } else {
      Status.CANCEL_STATUS
    }

}

case class CommandAnalysisEvent(val changedCommands: Set[Command], val docState: Document.State)

